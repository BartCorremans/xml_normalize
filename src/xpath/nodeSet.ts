import {XmlElement} from 'xmldoc';


export type NodeSet = XmlElement[];
export type ValueSet = string[];