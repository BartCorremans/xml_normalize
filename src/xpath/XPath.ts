import {parseXPathStep, XPathStep} from './XPathStep';

export interface XPath {
    absolute: boolean;
    steps: XPathStep[];
}

export function parseXPath(xpath: string): XPath {
    return {
        absolute: xpath.startsWith('/'),
        steps: xpath.split('/')
            .filter((step, i) => i > 0 || !!step) // skip first for absolute paths
            .map(step => parseXPathStep(step))
    };
}
