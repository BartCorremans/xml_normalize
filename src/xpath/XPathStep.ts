import {parseXPath, XPath} from './XPath';

export type Axis = 'attribute' | 'child' | 'parent' | 'self';

export type Predicate = number | PredicateFunction;

type FunctionName = 'starts-with' | 'contains';

export interface PredicateFunction {
    functionName: FunctionName;
    params: (string | XPath)[];
}

export interface XPathStep {
    axis: Axis;
    nodeTest?: string;
    predicate?: Predicate;
}

function parsePredicate(str: string | undefined): Predicate | undefined {
    if (str) {
        if (str.match(/^\d+$/)) {
            return parseInt(str);
        } else {
            const functionName = str.substring(0, str.indexOf('(')) as FunctionName;
            const paramsStr = str.substring(functionName.length + 1, str.length - 1);
            // assume params do not contain commas (","):
            const params = paramsStr.split(',').map(p => {
                const matchString = p.match(/^\s*"(.*)"\s*$/);
                if (matchString) {
                    return matchString[1];
                } else {
                    return parseXPath(p.trim());
                }
            });
            return {functionName, params};
        }
    } else {
        return undefined;
    }
}


export function parseXPathStep(step: string): XPathStep {
    const predicateStart = step.indexOf('[');
    const stepWithoutPredicate = predicateStart >= 0 ? step.substr(0, predicateStart) : step;
    const predicateString = predicateStart >= 0 ? step.substring(predicateStart + 1, step.length - 1) : undefined;
    let axis: Axis;
    let nodeTest: string | undefined;
    if (step.startsWith('@')) {
        axis = 'attribute';
        nodeTest = stepWithoutPredicate.substr(1);
    } else if (step.startsWith('..')) {
        axis = 'parent';
    } else if (step.startsWith('.')) {
        axis = 'self';
    } else {
        axis = 'child';
        nodeTest = stepWithoutPredicate;
    }

    const predicate = parsePredicate(predicateString);
    return {axis, nodeTest, predicate};
}