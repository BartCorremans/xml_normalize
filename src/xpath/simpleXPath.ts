import {XmlDocument, XmlElement, XmlNode} from 'xmldoc';
import {Predicate, XPathStep} from './XPathStep';
import {NodeSet, ValueSet} from './nodeSet';
import {parseXPath, XPath} from './XPath';


export class Evaluator {
    constructor(private readonly doc: XmlDocument) {
    }

    evalNodeSet(xpath: string, context: XmlElement = this.doc): NodeSet {
        return this.eval(context, parseXPath(xpath)) as NodeSet;
    }

    evalValues(xpath: string, context: XmlElement = this.doc): string[] {
        return this.evalValuesParsed(parseXPath(xpath), context);
    }

    private evalValuesParsed(path: XPath, context: XmlElement = this.doc): string[] {
        const result = (this.eval(context, path) as (XmlElement | string)[])
            .filter(v => v !== undefined);
        const first = result[0];
        if (typeof first === 'string' || first === undefined) {
            return result as string[];
        } else {
            return (result as NodeSet).map(n => n.val);
        }
    }

    private eval(context: XmlElement, path: XPath): NodeSet | ValueSet {
        let nodeSet: NodeSet | ValueSet = [context];
        path.steps.forEach((s, stepIndex) => {
                if (stepIndex === 0 && path.absolute) {
                    nodeSet = this.matchesNodeTest(context, s.nodeTest) && this.matchesPredicate(context, 0, s.predicate) ? [context] : [];
                } else if (stepIndex === path.steps.length - 1 && s.axis === 'attribute') {
                    nodeSet = (nodeSet as NodeSet).reduce((all, n, i) => {
                        return [...all, this.evalNodeSetStep(n, s) as string];
                    }, [] as ValueSet);
                } else {
                    nodeSet = (nodeSet as NodeSet).reduce((all, n, i) => {
                        return [...all, ...this.evalNodeSetStep(n, s) as NodeSet];
                    }, [] as NodeSet);
                }
            }
        );
        return nodeSet;
    }

    private evalNodeSetStep(n: XmlElement, s: XPathStep): NodeSet | string {
        switch (s.axis) {
            case 'self':
                return [n];
            case 'child':
                return (n.children
                    .filter(c => this.matchesNodeTest(c, s.nodeTest)) as XmlElement[])
                    .filter((c, i) => this.matchesPredicate(c, i, s.predicate));
            case 'attribute':
                return n.attr[s.nodeTest!];
        }
        throw new Error(`not supported: ${s.axis} ${(s.nodeTest)} ${s.predicate}`);
    }

    private matchesNodeTest(c: XmlNode, nodeTest: string | undefined) {
        return c.type === 'element' && (!nodeTest || nodeTest === '*' || nodeTest === c.name);
    }

    private matchesPredicate(c: XmlElement, index: number, predicate: Predicate | undefined): boolean {
        if (predicate === undefined) {
            return true;
        } else if (typeof predicate === 'number') {
            return index + 1 === predicate;
        } else {
            const paramsValues = predicate.params.map(p => typeof p === 'string' ? p : this.evalValuesParsed(p, c)[0]);
            switch (predicate.functionName) {
                case 'contains':
                    return paramsValues[0].includes(paramsValues[1]);
                case 'starts-with':
                    return paramsValues[0].startsWith(paramsValues[1]);
                default:
                    throw new Error(`not supported predicate: ${predicate}`);
            }
        }
    }

    * allChildren(el: XmlNode, includeEl = true): IterableIterator<XmlNode> {
        if (includeEl) {
            yield el;
        }
        if (el.type === 'element') {
            for (const c of el.children) {
                yield* this.allChildren(c, true);
            }
        }
    }


    getParent(r: XmlNode): XmlElement {
        // TODO use index?
        for (const n of this.allChildren(this.doc)) {
            if (n.type === 'element' && n.children.indexOf(r) >= 0) {
                return n;
            }
        }
        throw new Error('could not find parent for: ' + r);
    }

}