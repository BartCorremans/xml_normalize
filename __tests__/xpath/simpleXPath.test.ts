import {XmlDocument} from 'xmldoc';
import {Evaluator} from '../../src/xpath/simpleXPath';

describe('simpleXPath', () => {
    describe('Evaluator', () => {
        const doc = new XmlDocument('' +
            '<root>' +
            ' <subnode>' +
            '  <node id="1">text1</node>' +
            '  <node id="2">text2</node>' +
            ' </subnode>' +
            ' <subnode>' +
            '  <node id="3">text3</node>' +
            '  <node id="4">text4</node>' +
            ' </subnode>' +
            '</root>');

        const evaluator = new Evaluator(doc);

        describe('evalNodeSet', () => {
            it('should get root', () => {
                const r = evaluator.evalNodeSet('/root');
                expect(r).toEqual([doc]);
            });
            it('should get simple set', () => {
                const r = evaluator.evalNodeSet('/root/subnode/node[2]');
                expect(r).toHaveLength(2);
                expect(r).toMatchSnapshot();
            });
            it('should get wildcard', () => {
                const mixedDoc = new XmlDocument('' +
                    '<root>' +
                    ' <subnode>' +
                    '  <node id="1">text1</node>' +
                    '  <anotherNode id="2">text2</anotherNode>' +
                    ' </subnode>' +
                    ' <anotherSubnode>' +
                    '  <node id="3">text3</node>' +
                    '  <node id="4">text4</node>' +
                    ' </anotherSubnode>' +
                    '</root>');
                const mixedEvaluator = new Evaluator(mixedDoc);
                const r = mixedEvaluator.evalNodeSet('/root/*/node');
                expect(r).toHaveLength(3);
                expect(r).toMatchSnapshot();
            });
            it('should work with starts-with', () => {
                const mixedDoc = new XmlDocument('' +
                    '<root>' +
                    ' <subnode>' +
                    '  <node id="a1">text1</node>' +
                    '  <node id="a2">text2</node>' +
                    ' </subnode>' +
                    ' <anotherSubnode>' +
                    '  <node id="a3">text3</node>' +
                    '  <node id="Z4">text4</node>' +
                    ' </anotherSubnode>' +
                    '</root>');
                const mixedEvaluator = new Evaluator(mixedDoc);
                const r = mixedEvaluator.evalValues('/root/*/node[starts-with(@id,"a")]/@id');
                expect(r).toEqual(['a1', 'a2', 'a3']);
            });
        });
        describe('evalValue', () => {
            it('should evaluate simple values', () => {
                const r = evaluator.evalValues('/root/subnode/node/@id');
                expect(r).toEqual(['1', '2', '3', '4']);
            });
            it('should evaluate text values', () => {
                const r = evaluator.evalValues('/root/subnode[1]/node');
                expect(r).toEqual(['text1', 'text2']);
            });
            it('should omit empty values', () => {
                const mixedDoc = new XmlDocument('' +
                    '<root>' +
                    ' <subnode>' +
                    '  <node id="1">text1</node>' +
                    '  <anotherNode id="2">text2</anotherNode>' +
                    ' </subnode>' +
                    ' <anotherSubnode>' +
                    '  <node>text3</node>' +
                    '  <node id="4">text4</node>' +
                    ' </anotherSubnode>' +
                    '</root>');
                const mixedEvaluator = new Evaluator(mixedDoc);
                const r = mixedEvaluator.evalValues('/root/*/node/@id');
                expect(r).toEqual(['1', '4']);
            });
        });
    });
});