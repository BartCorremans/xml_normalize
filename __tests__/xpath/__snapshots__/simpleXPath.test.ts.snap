// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simpleXPath Evaluator evalNodeSet should get simple set 1`] = `
Array [
  XmlElement {
    "attr": Object {
      "id": "2",
    },
    "children": Array [
      XmlTextNode {
        "text": "text2",
      },
    ],
    "column": 58,
    "firstChild": XmlTextNode {
      "text": "text2",
    },
    "lastChild": XmlTextNode {
      "text": "text2",
    },
    "line": 0,
    "name": "node",
    "position": 58,
    "startTagPosition": 46,
    "val": "text2",
  },
  XmlElement {
    "attr": Object {
      "id": "4",
    },
    "children": Array [
      XmlTextNode {
        "text": "text4",
      },
    ],
    "column": 133,
    "firstChild": XmlTextNode {
      "text": "text4",
    },
    "lastChild": XmlTextNode {
      "text": "text4",
    },
    "line": 0,
    "name": "node",
    "position": 133,
    "startTagPosition": 121,
    "val": "text4",
  },
]
`;

exports[`simpleXPath Evaluator evalNodeSet should get wildcard 1`] = `
Array [
  XmlElement {
    "attr": Object {
      "id": "1",
    },
    "children": Array [
      XmlTextNode {
        "text": "text1",
      },
    ],
    "column": 31,
    "firstChild": XmlTextNode {
      "text": "text1",
    },
    "lastChild": XmlTextNode {
      "text": "text1",
    },
    "line": 0,
    "name": "node",
    "position": 31,
    "startTagPosition": 19,
    "val": "text1",
  },
  XmlElement {
    "attr": Object {
      "id": "3",
    },
    "children": Array [
      XmlTextNode {
        "text": "text3",
      },
    ],
    "column": 127,
    "firstChild": XmlTextNode {
      "text": "text3",
    },
    "lastChild": XmlTextNode {
      "text": "text3",
    },
    "line": 0,
    "name": "node",
    "position": 127,
    "startTagPosition": 115,
    "val": "text3",
  },
  XmlElement {
    "attr": Object {
      "id": "4",
    },
    "children": Array [
      XmlTextNode {
        "text": "text4",
      },
    ],
    "column": 154,
    "firstChild": XmlTextNode {
      "text": "text4",
    },
    "lastChild": XmlTextNode {
      "text": "text4",
    },
    "line": 0,
    "name": "node",
    "position": 154,
    "startTagPosition": 142,
    "val": "text4",
  },
]
`;
